src/dtls_combine.rs

```diff
@@ -0,0 +1,219 @@
+pub fn combine_dtls_fragments<'a>(
+    fragments: &[DTLSMessageHandshake],
+    out: &'a mut [u8],
+) -> IResult<&'a [u8], Option<DTLSMessage<'a>>> {
+    if fragments.is_empty() {
+        return Ok((&[], None));
+    }
+
+    if fragments.len() > MAX_FRAGMENTS {
+        return Err(Err::Error(make_error(&*out, ErrorKind::TooLarge)));
+    }
+
+    const MESSAGE_HEADER_OFFSET: usize = 12;
+
+    // The header all of the fragments share the same DTLSMessage start, apart from the
+    // fragment information. This goes into the first 12 bytes.
+    if out.len() < MESSAGE_HEADER_OFFSET {
+        // Error case 1
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    }
+
+    // Helper to iterate the fragments in order.
+    let ordered = Ordered::new(fragments);
+
+    // Investigate each fragment_offset + fragment_length to figure out
+    // the max contiguous range over the fragments.
+    let max = ordered.max_contiguous();
+
+    // Unwrap is OK, because we have at least one item (checked above).
+    let first_handshake = ordered.iter().next().unwrap();
+
+    if first_handshake.fragment_offset != 0 {
+        // The first fragment must start at 0, or we might have
+        // missing packets or arriving out of order.
+        return Ok((&[], None));
+    }
+
+    let msg_type = first_handshake.msg_type;
+    let message_seq = first_handshake.message_seq;
+    let length = first_handshake.length;
+
+    #[allow(clippy::comparison_chain)]
+    if max > length {
+        // Error case 4
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    } else if max < length {
+        // We do not have all fragments yet
+        return Ok((&[], None));
+    }
+
+    // Write the header into output.
+    {
+        out[0] = msg_type.into(); // The type.
+        out[1..4].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit length
+        out[4..6].copy_from_slice(&message_seq.to_be_bytes()); // 16 bit message sequence
+        out[6..9].copy_from_slice(&[0, 0, 0]); // 24 bit fragment_offset, which is 0 for the entire message.
+        out[9..12].copy_from_slice(&length.to_be_bytes()[1..]); // 24 bit fragment_length, which is entire length.
+    }
+
+    let data = &mut out[MESSAGE_HEADER_OFFSET..];
+
+    if data.len() < length as usize {
+        // Error case 1
+        return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+    }
+
+    // Loop the fragments, in order and output the data.
+    for handshake in ordered.iter() {
+        if msg_type != handshake.msg_type {
+            // Error case 2
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.length != length {
+            // Error case 3
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
+
+        if handshake.message_seq != message_seq {
+            // Error case 6
+            return Err(Err::Error(make_error(&*out, ErrorKind::Fail)));
+        }
```

Comment:
This method will return `Fail` for every possible error, which will make debugging complicated. Instead, this could justify returning a different error kind. Maybe the [`context`](https://docs.rs/nom/7.1.3/nom/error/fn.context.html) feature could help?

Reply:
replying to the `context` suggestion: this may not easy to use here. Maybe a custom error type would be better

Reply:
The usual method in `nom` is to create a custom error type, an implement `From<ErrorKind>` and `ParseError`. See [x509-parser errors](https://github.com/rusticata/x509-parser/blob/x509-parser-0.16.0/src/error.rs) for an example.

But, please wait before changing errors, because this will add code, and this can be done in a second step